### ISSUE TITLE
GCC6 Support

### ISSUE THREAD

**Issue by cinemast** [open-source-parsers/jsoncpp#411]
Hi!

At Debian we are currently preparing the migration to GCC6. The following issues are currently affecting libjsoncpp.

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=811958

Could you take a look at it?

**Comment by cdunn2001** [comment]
This weekend.

Any chance you can update to 1.6.x? The link refers to 0.10.5.

**Comment by cinemast** [comment]
Tha transition is currently in progress. It might take a few days until its completed.

https://release.debian.org/transitions/html/auto-libjsoncpp.html

So you think those problems should be fixed in 1.6.x?

**Comment by cinemast** [comment]
I tried rebuilding the latest release with GCC-6, resulting in the following build errors:

http://paste.debian.net/378673/

**Comment by cinemast** [comment]
Also while we are at it, could you take a look at current debian patches and check if any of those could be imported upstream?

http://anonscm.debian.org/cgit/collab-maint/libjsoncpp.git/tree/debian/patches

**Comment by cdunn2001** [comment]
> So you think those problems should be fixed in 1.6.x?

No, but the fix depends on compiler version. (`std::auto_ptr<>` -> `const std::scoped_ptr<>`) I'd rather not fix this at all in the old code. If we can fix the conversion errors from your latest build, that's a better path forward.

> Also while we are at it, could you take a look at current debian patches and check if any of those could be imported upstream?

Sure! This weekend.

**Comment by cinemast** [comment]
Of course, I am working on dropping 0.x.y support in the next Debian release. Thanks for taking the effort.

Greetings

**Comment by cdunn2001** [comment]
Patched 0.y.z and fixed 1.y.z. Let's try again.

**Comment by cinemast** [comment]
Hi Cristopher!
With the latest git head on master, the errors are still mostly the same:

```
mkdir build
root@cinemast-ThinkPad-T410s:/tmp/jsoncpp# cd build/
root@cinemast-ThinkPad-T410s:/tmp/jsoncpp/build# cmake ..
-- The C compiler identification is GNU 6.0.0
-- The CXX compiler identification is GNU 6.0.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- JsonCpp Version: 1.7.0
-- Could NOT find PythonInterp (missing:  PYTHON_EXECUTABLE) (Required is at least version "2.6")
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/jsoncpp/build
root@cinemast-ThinkPad-T410s:/tmp/jsoncpp/build# make -j4
Scanning dependencies of target jsoncpp_lib_static
[ 11%] Building CXX object src/lib_json/CMakeFiles/jsoncpp_lib_static.dir/json_reader.cpp.o
[ 33%] Building CXX object src/lib_json/CMakeFiles/jsoncpp_lib_static.dir/json_value.cpp.o
[ 33%] Building CXX object src/lib_json/CMakeFiles/jsoncpp_lib_static.dir/json_writer.cpp.o
/tmp/jsoncpp/src/lib_json/json_value.cpp: In copy constructor 'Json::Value::CZString::CZString(const Json::Value::CZString&)':
/tmp/jsoncpp/src/lib_json/json_value.cpp:235:18: error: conversion to 'unsigned char:2' from 'unsigned int' may alter its value [-Werror=conversion]
   storage_.policy_ = (other.cstr_
                      ~~~~~~~~~~~~
                  ? (static_cast<DuplicationPolicy>(other.storage_.policy_) == noDuplication
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                      ? noDuplication : duplicate)
                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                  : static_cast<DuplicationPolicy>(other.storage_.policy_));
                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

/tmp/jsoncpp/src/lib_json/json_value.cpp:235:18: error: conversion to 'unsigned char:2' from 'unsigned int' may alter its value [-Werror=conversion]
/tmp/jsoncpp/src/lib_json/json_value.cpp: In instantiation of 'bool Json::InRange(double, T, U) [with T = long long int; U = long long int]':
/tmp/jsoncpp/src/lib_json/json_value.cpp:706:51:   required from here
/tmp/jsoncpp/src/lib_json/json_value.cpp:61:12: error: conversion to 'double' from 'long long int' may alter its value [-Werror=conversion]
   return d >= min && d <= max;
          ~~^~~~~~

/tmp/jsoncpp/src/lib_json/json_value.cpp:61:24: error: conversion to 'double' from 'long long int' may alter its value [-Werror=conversion]
   return d >= min && d <= max;
                      ~~^~~~~~

/tmp/jsoncpp/src/lib_json/json_value.cpp: In instantiation of 'bool Json::InRange(double, T, U) [with T = int; U = long long unsigned int]':
/tmp/jsoncpp/src/lib_json/json_value.cpp:727:45:   required from here
/tmp/jsoncpp/src/lib_json/json_value.cpp:61:24: error: conversion to 'double' from 'long long unsigned int' may alter its value [-Werror=conversion]
cc1plus: some warnings being treated as errors
src/lib_json/CMakeFiles/jsoncpp_lib_static.dir/build.make:86: recipe for target 'src/lib_json/CMakeFiles/jsoncpp_lib_static.dir/json_value.cpp.o' failed
make[2]: *** [src/lib_json/CMakeFiles/jsoncpp_lib_static.dir/json_value.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
/tmp/jsoncpp/src/lib_json/json_writer.cpp: In member function 'bool Json::BuiltStyledStreamWriter::isMultineArray(const Json::Value&)':
/tmp/jsoncpp/src/lib_json/json_writer.cpp:1014:31: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
   bool isMultiLine = size * 3 >= rightMargin_;
                      ~~~~~~~~~^~~~~~~~~~~~~~~

/tmp/jsoncpp/src/lib_json/json_writer.cpp:1034:45: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     isMultiLine = isMultiLine || lineLength >= rightMargin_;
                                  ~~~~~~~~~~~^~~~~~~~~~~~~~~

CMakeFiles/Makefile2:109: recipe for target 'src/lib_json/CMakeFiles/jsoncpp_lib_static.dir/all' failed
make[1]: *** [src/lib_json/CMakeFiles/jsoncpp_lib_static.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by cinemast** [comment]
Do you have a GCC6 installation available somewhere? GCC6 is available in Debian experimental.

**Comment by cdunn2001** [comment]
I don't have easy access to gcc-6. However, I was able to produce all but 1 of those errors via `-Werror=xxx`, and I think I've fixed the last one too. See #416.

**Comment by cdunn2001** [comment]
Did those fixes work?

**Comment by cinemast** [comment]
Hi Christopher!

Only partly, this one is still a problem:

```
/tmp/jsoncpp/src/lib_json/json_value.cpp:235:18: error: conversion to 'unsigned char:2' from 'unsigned int' may alter its value [-Werror=conversion]
   storage_.policy_ = (other.cstr_
                      ~~~~~~~~~~~~
                  ? (static_cast<DuplicationPolicy>(other.storage_.policy_) == noDuplication
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                      ? noDuplication : duplicate)
                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                  : static_cast<DuplicationPolicy>(other.storage_.policy_));
```

**Comment by cdunn2001** [comment]
Try again?

Or can you suggest a fix? We definitely need to assign an `unsigned int` into a 2-bit unsigned field.

**Comment by cdunn2001** [comment]
@cinemast, try again. @ya1gaurav has tested the tip of master.

**Comment by cinemast** [comment]
Great! I can confirm it works.

**Comment by ya1gaurav** [comment]
@cinemast Do you face GGC 6 issues in some other open source also ?

**Comment by cinemast** [comment]
What exactly do you mean by that? In Debian there are currently lots of compile errors due to GCC6. 

https://bugs.debian.org/cgi-bin/pkgreport.cgi?tag=ftbfs-gcc-6;users=debian-gcc@lists.debian.org

**Comment by ya1gaurav** [comment]
I mean I have setup for GCC 6.0, so I can try solving build issues related to GCC 6.0 in other open source softwares also.

**Comment by ya1gaurav** [comment]
And now it seems - https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=811958  got fixed.

**Comment by cinemast** [comment]
Exactly :) I will just wait for the next release instead of patching manually.

**Comment by ya1gaurav** [comment]
@cdunn2001 Could you help @cinemast with new release featuring GCC 6.0 support.

**Comment by cdunn2001** [comment]
Released and signed `1.7.0`.
